Series of minor deflate optimizations/cleanups

Dead2 opened the pull request:
- Don't check for too long match in deflate_quick, it cannot happen.
- Minor inlining change in trees_emit.h
      - Inline the small bi_windup function
      - Don't attempt inlining for the big zng_emit_dist
- Reduce opt_len/static_len sizes.
- Add local window pointer to:
  - deflate_quick
  - deflate_fast
  - deflate_medium
  - fill_window
- Add local strm pointer in fill_window.
- Fix missed change to use local lookahead variable in match_tpl
- Move matches/insert in deflate_state closer to their related varibles.
   These also fill a 8-byte hole in the struct on 64-bit platforms.
- Exclude compressed_len and bits_sent if ZLIB_DEBUG is not enabled. Also move them to the end.
- Add local window pointer to deflate_quick, deflate_fast and deflate_medium.
- Reorder variables in longest_match, reducing gaps.
- Move GOTO_NEXT_CHAIN macro outside of longest_match function to improve readability.
- Make window-based pointers in match_tpl.h const, only the pointers move, never the data.
- Use pointer arithmetic to access window in deflate_quick/deflate_fast.
- Use uint32_t for Pos in match_tpl.h
- Use uint32_t for hash_head in update_hash/insert_string
- Reorder deflate.h variables to improve cache locality

Compression performance on x86_64 is about 5-6% faster.
On aarch64 it is around 1.3% faster.

Dead2 commented:
### Develop
 ```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.185%  0.0577/0.0579/0.0580/0.0000    0.0241/0.0243/0.0243/0.0001        8,526,745
 2     43.871%  0.1017/0.1019/0.1021/0.0001    0.0239/0.0240/0.0240/0.0000        6,903,702
 3     42.388%  0.1231/0.1234/0.1235/0.0001    0.0230/0.0231/0.0232/0.0000        6,670,239
 4     41.647%  0.1499/0.1508/0.1509/0.0002    0.0224/0.0225/0.0226/0.0001        6,553,746
 5     41.216%  0.1673/0.1674/0.1675/0.0000    0.0224/0.0225/0.0225/0.0000        6,485,936
 6     41.038%  0.2152/0.2163/0.2164/0.0002    0.0220/0.0221/0.0221/0.0000        6,457,827
 7     40.778%  0.3158/0.3163/0.3164/0.0001    0.0223/0.0225/0.0225/0.0000        6,416,941
 8     40.704%  0.4169/0.4177/0.4183/0.0004    0.0223/0.0224/0.0224/0.0000        6,405,249
 9     40.409%  0.4542/0.4549/0.4551/0.0002    0.0217/0.0218/0.0218/0.0000        6,358,951

 avg1  42.915%                       0.2230                         0.0228
 tot                                80.2653                         8.2022       60,779,336

   text    data     bss     dec     hex filename
 168756    1352       8  170116   29884 libz-ng.so.2
  ```
  
 ### PR
  ```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.185%  0.0562/0.0568/0.0569/0.0001    0.0237/0.0244/0.0245/0.0001        8,526,745
 2     43.871%  0.0992/0.0996/0.0998/0.0002    0.0240/0.0241/0.0241/0.0000        6,903,702
 3     42.388%  0.1220/0.1224/0.1227/0.0002    0.0230/0.0232/0.0232/0.0000        6,670,239
 4     41.647%  0.1496/0.1503/0.1507/0.0003    0.0226/0.0227/0.0227/0.0000        6,553,746
 5     41.216%  0.1666/0.1667/0.1668/0.0001    0.0225/0.0226/0.0226/0.0000        6,485,936
 6     41.038%  0.2145/0.2147/0.2148/0.0001    0.0221/0.0222/0.0222/0.0000        6,457,827
 7     40.778%  0.3134/0.3139/0.3140/0.0001    0.0225/0.0226/0.0226/0.0000        6,416,941
 8     40.704%  0.4146/0.4155/0.4161/0.0004    0.0225/0.0225/0.0226/0.0000        6,405,249
 9     40.409%  0.4546/0.4550/0.4552/0.0002    0.0218/0.0219/0.0219/0.0000        6,358,951

 avg1  42.915%                       0.2217                         0.0229
 tot                                79.7969                         8.2450       60,779,336

   text    data     bss     dec     hex filename
 168956    1352       8  170316   2994c libz-ng.so.2
 ```
 
 200 bytes bigger code size.
 0.58% faster compression overall
 Level 6 is 0.74% faster

nmoinvaz commented on match_tpl.h:
```diff
@@ -24,33 +29,31 @@
     /* The code is optimized for STD_MAX_MATCH-2 multiple of 16. */
     Assert(STD_MAX_MATCH == 258, "Code too clever");
 
+    scan = window + strstart;
```
Why not assign this during variable declaration?

Dead2 replied:
strstart is not set when *scan is defined above. Reordering them could fix that, but due the to differing alignment requirements of pointer and int, it'd require moving two variables, and make it harder to keep the alignment intact in the future. I think the compiler is smart enough to do the assignment early anyway.

nmoinvaz reviewed (commented):
I think first commit might need to be split up. I made one comment. Otherwise looks ok to me.

Dead2 committed:
- Minor inlining changes in trees_emit.h:
  - Inline the small bi_windup function
  - Don't attempt inlining for the big zng_emit_dist
- Don't check for too long match in deflate_quick, it cannot happen.
- Move GOTO_NEXT_CHAIN macro outside of LONGEST_MATCH function to
  improve readability.

Dead2 committed:
Deflate_state changes:
- Reduce opt_len/static_len sizes.
- Move matches/insert closer to their related varibles.
  These now fill a 8-byte hole in the struct on 64-bit platforms.
- Exclude compressed_len and bits_sent if ZLIB_DEBUG is
  not enabled. Also move them to the end.
- Remove x86 MSVC-specific padding

Dead2 committed:
- Add local window pointer to:
  deflate_quick, deflate_fast, deflate_medium and fill_window.
- Add local strm pointer in fill_window.
- Fix missed change to use local lookahead variable in match_tpl

Dead2 commented:
> I think first commit might need to be split up. I made one comment. Otherwise looks ok to me.

I remastered the commits to group the changes by type now, making reviewing easier.

mtl1979 commented on deflate_fast.c:
```diff
@@ -44,9 +45,9 @@ Z_INTERNAL block_state deflate_fast(deflate_state *s, int flush) {
         if (s->lookahead >= WANT_MIN_MATCH) {
 #if BYTE_ORDER == LITTLE_ENDIAN
-            uint32_t str_val = zng_memread_4(&s->window[s->strstart]);
+            uint32_t str_val = zng_memread_4(&window[s->strstart]);
```
This can be written as "window + s->strstart", which removes one indirection and other places already use pointer arithmetic in function parameters.

Dead2 replied:
implemented

mtl1979 commented on deflate_fast.c:
```diff
@@ -44,9 +45,9 @@ Z_INTERNAL block_state deflate_fast(deflate_state *s, int flush) {
+            uint32_t str_val = zng_memread_4(&window[s->strstart]);
 #else
-            uint32_t str_val = ZSWAP32(zng_memread_4(&s->window[s->strstart]));
+            uint32_t str_val = ZSWAP32(zng_memread_4(&window[s->strstart]));
```
Same as above.

mtl1979 commented on deflate_quick.c:
```diff
@@ -85,32 +89,31 @@ Z_INTERNAL block_state deflate_quick(deflate_state *s, int flush) {
         if (LIKELY(s->lookahead >= WANT_MIN_MATCH)) {
 #if BYTE_ORDER == LITTLE_ENDIAN
-            uint32_t str_val = zng_memread_4(&s->window[s->strstart]);
+            uint32_t str_val = zng_memread_4(&window[s->strstart]);
```
"window + s->strstart" is a lot more readable and pointer arithmetics are already used elsewhere.

mtl1979 commented on deflate_quick.c:
```diff
@@ -85,32 +89,31 @@ Z_INTERNAL block_state deflate_quick(deflate_state *s, int flush) {
+            uint32_t str_val = zng_memread_4(&window[s->strstart]);
 #else
-            uint32_t str_val = ZSWAP32(zng_memread_4(&s->window[s->strstart]));
+            uint32_t str_val = ZSWAP32(zng_memread_4(&window[s->strstart]));
```
Same as above...

mtl1979 reviewed (changes requested):
Inconsistent style in some edited files. Further cleanup possible.

Dead2 committed:
Use pointer arithmetic to access window in deflate_quick/deflate_fast

Dead2 committed:
- Reorder variables in longest_match, reducing gaps.
- Make window-based pointers in match_tpl.h const, only the
  pointers move, never the data.

Dead2 commented on deflate_quick.c:
```diff
@@ -85,32 +89,31 @@ Z_INTERNAL block_state deflate_quick(deflate_state *s, int flush) {
-                        if (UNLIKELY(match_len > STD_MAX_MATCH))
-                            match_len = STD_MAX_MATCH;
 
+                        Assert(match_len <= STD_MAX_MATCH, "match too long");
```
Calls to compare256() can only return values up to 256, thus match_len will at most be 258.

Dead2 committed:
Reorder deflate.h variables to improve cache locality

Dead2 committed:
Use uin32_t for Pos in match_tpl.h

Dead2 committed:
Use uint32_t for hash_head in update_hash/insert_string

Dead2 commented:
i7-11700K
### Develop 9 Dec
 ```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.164%  0.1963/0.1969/0.1973/0.0002    0.0791/0.0793/0.0794/0.0001       21,937,619
 2     39.504%  0.3198/0.3207/0.3217/0.0005    0.0780/0.0787/0.0790/0.0004       17,993,257
 3     37.948%  0.3946/0.3955/0.3962/0.0004    0.0740/0.0742/0.0742/0.0000       17,284,459
 4     37.258%  0.4558/0.4570/0.4582/0.0005    0.0722/0.0723/0.0724/0.0001       16,969,880
 5     36.777%  0.5015/0.5025/0.5038/0.0006    0.0717/0.0725/0.0727/0.0003       16,750,917
 6     36.633%  0.6092/0.6106/0.6119/0.0006    0.0713/0.0722/0.0723/0.0003       16,685,504
 7     36.389%  0.8560/0.8570/0.8579/0.0005    0.0720/0.0729/0.0730/0.0002       16,574,444
 8     36.295%  1.3075/1.3086/1.3099/0.0006    0.0728/0.0729/0.0730/0.0000       16,531,506
 9     36.114%  1.4002/1.4018/1.4026/0.0006    0.0701/0.0703/0.0703/0.0001       16,449,018

 avg1  38.343%                       0.6723                         0.0739
 tot                               405.3834                        44.5683      157,176,604

   text    data     bss     dec     hex filename
 169756    1360       8  171124   29c74 libz-ng.so.2
  ```
  
  ### PR
  ```
  Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.164%  0.1983/0.1989/0.1994/0.0003    0.0791/0.0793/0.0794/0.0001       21,937,619
 2     39.504%  0.3106/0.3116/0.3127/0.0006    0.0782/0.0789/0.0791/0.0003       17,993,257
 3     37.948%  0.3875/0.3889/0.3902/0.0007    0.0741/0.0742/0.0743/0.0000       17,284,459
 4     37.258%  0.4359/0.4372/0.4381/0.0006    0.0722/0.0723/0.0724/0.0000       16,969,880
 5     36.777%  0.4819/0.4828/0.4839/0.0005    0.0718/0.0720/0.0727/0.0002       16,750,917
 6     36.633%  0.5784/0.5794/0.5804/0.0005    0.0714/0.0720/0.0724/0.0004       16,685,504
 7     36.389%  0.7843/0.7856/0.7868/0.0007    0.0721/0.0730/0.0731/0.0002       16,574,444
 8     36.295%  1.1744/1.1755/1.1768/0.0006    0.0720/0.0729/0.0730/0.0003       16,531,506
 9     36.114%  1.3142/1.3155/1.3164/0.0006    0.0694/0.0703/0.0704/0.0002       16,449,018

 avg1  38.343%                       0.6306                         0.0739
 tot                               380.2512                        44.5431      157,176,604

   text    data     bss     dec     hex filename
 169396    1360       8  170764   29b0c libz-ng.so.2
 ```
 360 bytes smaller
 6.2% faster compression on average.
 Level 6 is 5.1% faster.

Dead2 commented:
AMD 8700GE Zen4
### Develop Dec 9
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.164%  0.1337/0.1346/0.1349/0.0004    0.0566/0.0568/0.0570/0.0001       21,937,619
 2     39.504%  0.2350/0.2357/0.2360/0.0003    0.0557/0.0560/0.0565/0.0002       17,993,257
 3     37.948%  0.2869/0.2877/0.2880/0.0003    0.0530/0.0533/0.0537/0.0002       17,284,459
 4     37.258%  0.3458/0.3469/0.3477/0.0004    0.0517/0.0520/0.0524/0.0002       16,969,880
 5     36.777%  0.3882/0.3891/0.3897/0.0004    0.0514/0.0517/0.0521/0.0002       16,750,917
 6     36.633%  0.4887/0.4895/0.4901/0.0004    0.0510/0.0513/0.0517/0.0002       16,685,504
 7     36.389%  0.7004/0.7014/0.7019/0.0004    0.0514/0.0518/0.0522/0.0002       16,574,444
 8     36.295%  1.0681/1.0700/1.0713/0.0009    0.0512/0.0516/0.0520/0.0002       16,531,506
 9     36.114%  1.1277/1.1300/1.1307/0.0007    0.0499/0.0503/0.0507/0.0002       16,449,018

 avg1  38.343%                       0.5317                         0.0528
 tot                               320.5894                        31.8149      157,176,604
 
    text    data     bss     dec     hex filename
 155420    1352       8  156780   2646c libz-ng.so.2
```

# PR
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.164%  0.1295/0.1302/0.1304/0.0001    0.0569/0.0570/0.0571/0.0001       21,937,619
 2     39.504%  0.2272/0.2281/0.2284/0.0003    0.0558/0.0559/0.0560/0.0001       17,993,257
 3     37.948%  0.2767/0.2775/0.2779/0.0003    0.0530/0.0531/0.0532/0.0000       17,284,459
 4     37.258%  0.3265/0.3272/0.3275/0.0002    0.0517/0.0519/0.0519/0.0001       16,969,880
 5     36.777%  0.3696/0.3703/0.3708/0.0003    0.0514/0.0515/0.0516/0.0001       16,750,917
 6     36.633%  0.4635/0.4642/0.4647/0.0003    0.0511/0.0512/0.0513/0.0001       16,685,504
 7     36.389%  0.6587/0.6595/0.6599/0.0003    0.0514/0.0516/0.0517/0.0001       16,574,444
 8     36.295%  0.9895/0.9912/0.9924/0.0006    0.0514/0.0516/0.0517/0.0001       16,531,506
 9     36.114%  1.0625/1.0632/1.0636/0.0003    0.0499/0.0501/0.0502/0.0001       16,449,018

 avg1  38.343%                       0.5013                         0.0527
 tot                               302.2631                        31.7588      157,176,604

   text    data     bss     dec     hex filename
 155252    1352       8  156612   263c4 libz-ng.so.2
 ```
 168 bytes smaller
 5.7% faster on average
Level 6 is 5.2% faster

Dead2 commented:
RPI5 Aarch64
### Develop Dec9
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.164%  0.2779/0.2901/0.2940/0.0043    0.0982/0.1126/0.1178/0.0053       21,937,619
 2     39.504%  0.4891/0.5038/0.5083/0.0044    0.0979/0.1092/0.1151/0.0048       17,993,257
 3     37.948%  0.5918/0.6071/0.6117/0.0046    0.0961/0.1052/0.1111/0.0046       17,284,459
 4     37.258%  0.6754/0.6822/0.6877/0.0038    0.0911/0.1024/0.1076/0.0046       16,969,880
 5     36.777%  0.7213/0.7342/0.7413/0.0048    0.0777/0.1011/0.1068/0.0053       16,750,917
 6     36.633%  0.8515/0.8621/0.8668/0.0040    0.0851/0.1013/0.1066/0.0053       16,685,504
 7     36.389%  1.1828/1.1965/1.2017/0.0054    0.0812/0.0991/0.1056/0.0055       16,574,444
 8     36.295%  1.7671/1.7769/1.7808/0.0041    0.0852/0.0999/0.1087/0.0053       16,531,506
 9     36.114%  1.9322/1.9435/1.9481/0.0042    0.0853/0.0997/0.1068/0.0048       16,449,018

 avg1  38.343%                       0.9552                         0.1034
 tot                               292.2812                        31.6368      157,176,604

   text    data     bss     dec     hex filename
 119189    1528       8  120725   1d795 libz-ng.so.2
```
 
### deflate_opt PR updated
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.164%  0.2759/0.2877/0.2916/0.0033    0.0983/0.1119/0.1177/0.0049       21,937,619
 2     39.504%  0.4869/0.4954/0.4996/0.0042    0.0940/0.1080/0.1138/0.0055       17,993,257
 3     37.948%  0.5849/0.5933/0.5967/0.0032    0.0917/0.1054/0.1107/0.0045       17,284,459
 4     37.258%  0.6526/0.6667/0.6724/0.0042    0.0924/0.1020/0.1083/0.0042       16,969,880
 5     36.777%  0.7112/0.7257/0.7298/0.0034    0.0833/0.0988/0.1066/0.0066       16,750,917
 6     36.633%  0.8407/0.8512/0.8567/0.0043    0.0829/0.1025/0.1074/0.0053       16,685,504
 7     36.389%  1.1687/1.1799/1.1858/0.0043    0.0892/0.1009/0.1068/0.0045       16,574,444
 8     36.295%  1.7568/1.7653/1.7688/0.0028    0.0835/0.0994/0.1063/0.0057       16,531,506
 9     36.114%  1.9025/1.9157/1.9222/0.0052    0.0881/0.0978/0.1044/0.0043       16,449,018

 avg1  38.343%                       0.9423                         0.1030
 tot                               288.3548                        31.5113      157,176,604

   text    data     bss     dec     hex filename
 119341    1528       8  120877   1d82d libz-ng.so.2
 ```
 152 bytes bigger
 1.35% faster on average
 Level 6 is 1.3% faster

mtl1979 commented on deflate.h:
```diff
@@ -221,16 +222,27 @@ struct ALIGNED_(64) internal_state {
 #if defined(_M_IX86) || defined(_M_ARM)
-    int padding[2];
+    int32_t padding2[1];
 #endif
```
Padding at start of cache line seems wrong... Padding should always be before alignment boundary.

Dead2 replied:
Where this buffer lands on win32 builds I do not know, the cacheline comments are gathered from linux x86_64, but are probably accurate for most 64-bit systems..
crc_fold is going away in one of @nmoinvaz PRs, so both of these paddings would probably then become unnecessary and could go away along with it.

mtl1979 replied:
As far as I understand 32-bit architectures, those paddings are supposed to be after for example pointers to align before any struct members that are wider than 32 bits... It usually helps to put the struct members on a spreadsheet that contain bit widths of all the types for both 32-bit and 64-bit architectures.

Temporarily removing the padding should cause compiler to emit warning around the struct member that needs the padding. We have most warnings enabled and treated as errors on both Win32 and Win64 MSVC

If I ask AI how many bytes there are to start of next cache line without any explicit padding, it says there is 8 bytes. So this would mean that the second padding (only for 32-bit architectures) must be placed before comment for start of cache line 3.

mtl1979 replied:
Before the struct definition:

```
#ifdef _MSC_VER
#pragma warning(push)
#pragma warning(error : 4820)
#endif
```

and after the definition:

```
#ifdef _MSC_VER
#pragma warning(pop)
#endif
```

That should break the build if there is padding mismatch...

Dead2 replied:
It would make more sense to fix the 32-bit padding after all the changes to the struct is done though.

nmoinvaz replied:
Yeah there is no point to try to fix it now.

mtl1979 replied:
Just moving the padding above the comment is enough for this PR... I found out that the padding need to be fixed on boundary of almost every cache line and inside at least one too... I have WIP commit for fixing them waiting for any other changes affecting the same struct...

mtl1979 reviewed (changes requested):
Padding on wrong side of alignment boundary, might introduce implicit padding we are trying to avoid.

nmoinvaz commented on deflate.c:
```diff
@@ -1171,7 +1171,9 @@ static void lm_init(deflate_state *s) {
  */
 
 void Z_INTERNAL PREFIX(fill_window)(deflate_state *s) {
+    PREFIX3(stream)      *strm = s->strm;
```
Fix whitespace.

nmoinvaz reviewed (approved):
LGTM
